codegeex.el

```diff
@@ -163,8 +244,126 @@ The result will be replaced into the selected region."
     (codegeex-debug-invoke prompt (codegeex-language) begin end))
   nil)
 
+;; minor mode
+
+(defvar codegeex--post-command-timer nil)
+
+(defcustom codegeex-disable-predicates nil
+  "A list of predicate functions with no argument to disable Codegeex.
+Codegeex will not be triggered if any predicate returns t."
+  :type '(repeat function)
+  :group 'codegeex)
+
+(defcustom codegeex-enable-predicates '(evil-insert-state-p codegeex--buffer-changed)
+  "A list of predicate functions with no argument to enable Codegeex.
+Codegeex will be triggered only if all predicates return t."
+  :type '(repeat function)
+  :group 'codegeex)
+
+(defcustom codegeex-disable-display-predicates nil
+  "A list of predicate functions with no argument to disable Codegeex.
+Codegeex will not show completions if any predicate returns t."
+  :type '(repeat function)
+  :group 'codegeex)
+
+(defcustom codegeex-enable-display-predicates nil
+  "A list of predicate functions with no argument to enable Codegeex.
+Codegeex will show completions only if all predicates return t."
+  :type '(repeat function)
+  :group 'codegeex)
+
+(defmacro codegeex--satisfy-predicates (enable disable)
+  "Return t if satisfy all predicates in ENABLE and none in DISABLE."
+  `(and (cl-every (lambda (pred)
+                    (if (functionp pred) (funcall pred) t))
+                  ,enable)
+        (cl-notany (lambda (pred)
+                     (if (functionp pred) (funcall pred) nil))
+                   ,disable)))
```

Comment:
This is used to trigger the request to codegeex api. Maybe we can add options to avoid triggering it on a comment line or something. Or even using a *context* system where different api endpoint are used based on where the pointer is in the buffer.